scheduler/conflicts_checker.py

```diff
@@ -1,70 +1,93 @@
+import copy
 from typing import List, Tuple
 from enum import Enum
-from scheduler.models import Lesson, Professor, Auditorium, Group
 
-class ConflictType(Enum):
-    PROFESSOR = 1
-    AUDITORIUM = 2
-    GROUP = 3
+from django.db.models.query import QuerySet
+
+from scheduler.models import Lesson, Professor, Auditorium, Group, Conflict
+
 
 def are_overlapping(lesson1: Lesson, lesson2: Lesson) -> bool:
     """
     Check if one lesson started or ended during other
-    :param lesson1, lesson2: Lesson object retreived from database
+    :param lesson1: Lesson object retrieved from database
+    :param lesson2: Lesson object retrieved from database
     :return: bool if lessons are conflicting
     """
-    if lesson1.start_time <= lesson2.start_time and lesson1.end_time > lesson2.start_time:
+    if lesson1.start_time <= lesson2.start_time < lesson1.end_time:
         return True
-    if lesson1.end_time >= lesson2.end_time and lesson1.start_time < lesson2.end_time:
+    if lesson1.end_time >= lesson2.end_time > lesson1.start_time:
         return True
     if lesson1.start_time >= lesson2.start_time and lesson1.end_time <= lesson2.end_time:
         return True
     return False
 
-def check_lesson(lesson: Lesson) -> List[Tuple[ConflictType, Lesson, Lesson, object]]:
+
+def check_lesson(index: int, lesson_list: List[Lesson]) -> List[Tuple[str, Lesson, Lesson, int]]:
     """
     This function searches db for other lessons sharing the same professor, auditorium and group
-    And then checks if they are conflcting with eachother using are_overlapping
-    :param lesson: Lesson for which we are chcecking possible conflicts
+    And then checks if they are conflicting with each other using are_overlapping
+    :param index: index of Lesson for which we are checking possible conflicts
+    :param lesson_list: lessons list
     :return: List of tuples holding information about conflict.
-             Tuple[ConflictType, Lesson1, Lesson2, object]
-             ConflictType helps to differentiate which conflict it is
+             Tuple[str, Lesson1, Lesson2, int]
+             str helps to differentiate which conflict it is
              Lesson1 and Lesson2 are Lesson that are currently in conflict
-             object is Model object responsible for conflict (Professor,Auditorium,Group)
+             int is Model object id responsible for conflict (Professor,Auditorium,Group)
     """
-    conflicts: List[Tuple[ConflictType, Lesson, Lesson, object]] = []
-    for lesson_2 in Lesson.objects.filter(professor=lesson.professor):
-        if lesson != lesson_2:
-            if are_overlapping(lesson, lesson_2):
-                new_conflict = (ConflictType.PROFESSOR, lesson, lesson_2, lesson.professor)
+    conflicts: List[Tuple[str, Lesson, Lesson, int]] = []
+    lesson = lesson_list[index]
+    for ind in range(index, len(lesson_list)):
+        lesson_2 = lesson_list[ind]
+        if lesson != lesson_2 and are_overlapping(lesson, lesson_2):
```

Comment:
Skoro i tak zaczynamy od danej lekcji, to lepiej przesunąć o 1 index w range i w ifie sprawdzać tylko jeden warunek (are_overlapping)

Reply:
Słusznie :D